### DISCUSSION TITLE
How can I disable the onboard green heartbeat LED?

### DISCUSSION THREAD

**Discussion by nobara256** [Joshua-Riek/ubuntu-rockchip#156] [Q&A] [answered]
It seems like the green led is beating again with v1.15. As far I understand it's rhythm is related with the CPU frequency. It was silent in few previous builds and I would like to see it disabled again.

**Comment by Joshua-Riek** [comment]
There were a lot of changes to the kernel and device tree in v1.15. Normally something like this can be easily disabled in sysfs or the device tree, ill take a look later today.

**Comment by Joshua-Riek** [comment, marked as answer]
You can disable the led through the sysfs interface. The below command will need to be issued every boot, this can be done by putting the command in the file /etc/rc.local.
```bash
echo none > /sys/class/leds/status_led/trigger
```

**Reply by nobara256** [reply]
Applied and it works. Although the LED does blink for a few seconds during boot. Thank you so much!

**Reply by marcoexpo** [reply]
Hi, I have a Orange pi 5 pro and the command works for me. However, I can't find the /etc/rc.local file to edit it, so on every boot the LED starts blinking again. Can someone please help me with this. Thank you

**Reply by nobara256** [reply]
The best solution I found is to disable it in the device tree itself (it won't blink, not even during the boot).
If you are running kernel 6.1 (24.04 builds), then you can just add the overlay files ("disable-led" something) to the `/etc/default/u-boot` under the overlay field.
If you are running the mainline 6.11 (24.10 next builds) then you need to edit the device tree itself and change the trigger from "heartbeat" to "none".

**Reply by marcoexpo** [reply]
Soory, could you be more precise. I have kernel 6.1 (24.04 builds) installed. What do you mean by "overlay files"? My /etc/default/u-boot file looks like this:

## /etc/default/u-boot - configuration file for u-boot-update(8)

#U_BOOT_UPDATE="true"

#U_BOOT_ALTERNATIVES="default recovery"
#U_BOOT_DEFAULT="l0"
#U_BOOT_PROMPT="1"
#U_BOOT_ENTRIES="all"
#U_BOOT_MENU_LABEL="Debian GNU/Linux"
#U_BOOT_PARAMETERS="ro earlycon"
#U_BOOT_ROOT=""
#U_BOOT_TIMEOUT="50"
#U_BOOT_FDT=""
#U_BOOT_FDT_DIR="/lib/firmware/"
#U_BOOT_FDT_OVERLAYS=""
#U_BOOT_FDT_OVERLAYS_DIR="/lib/firmware/"
#U_BOOT_SYNC_DTBS="false"

Thank you again for the help!

**Reply by nobara256** [reply]
Sorry for the late reply.

Adding `device-tree/rockchip/overlay/orangepi-5-disable-leds.dtb` to `U_BOOT_FDT_OVERLAYS=""` should do the job. Remove `#` from that line.

Then you need to execute `sudo u-boot-update` and reboot to make the changes effective.

**Comment by SilverDesert** [comment]
Do you remove the exit 0 from the file or leave it as it is?

**Reply by Joshua-Riek** [reply]
Leave the exit 0 at the end of the file

**Comment by SilverDesert** [comment]
I tried it but it is still blinking. I rebooted and it is still blinking.

**Reply by Joshua-Riek** [reply]
Which board are you using?

**Comment by SilverDesert** [comment]
Orange Pi 5 Plus using V1.18

**Reply by Joshua-Riek** [reply]
Oh, there is a more elegant solution for the Orange Pi 5 Plus. A device tree overlay now exists to disable the LED. The overlay name is `orangepi-5-plus-disable-leds` which will need to be added to `ubuntuEnv.txt`. You can use the below command to do this, then give your system a reboot.

```bash
sudo sed -i 's/overlays=/overlays=orangepi-5-plus-disable-leds /g' /boot/firmware/ubuntuEnv.txt
```

**Reply by SilverDesert** [reply]
Thanks, it works perfectly. How do I revert if I want to revert the blue blinking led?

**Reply by SilverDesert** [reply]
Do I just replace 'enable' with the 'disable' part?

**Reply by Joshua-Riek** [reply]
A config file can be found at `/boot/firmware/ubuntuEnv.txt` and will look something like this.
```
bootargs=root=UUID=2370087c-25a4-4eaf-baea-1e24f3194383 rootfstype=ext4 rootwait rw console=ttyS2,1500000 console=tty1 cgroup_enable=cpuset cgroup_memory=1 cgroup_enable=memory swapaccount=1 systemd.unified_cgroup_hierarchy=0 quiet splash plymouth.ignore-serial-consoles
fdtfile=rk3588-orangepi-5-plus.dtb
overlay_prefix=orangepi-5-plus
overlays=orangepi-5-plus-disable-leds
```
You will need to remove the `orangepi-5-plus-disable-leds` text on the line that starts with `overlays=`.

**Reply by SilverDesert** [reply]
Is it possible to make the led blink meaningfully like it is with Raspberry Pi instead of just pulsating meaninglessly?

**Reply by Joshua-Riek** [reply]
Hmm, one moment, let me look at the device tree.

**Reply by Joshua-Riek** [reply]
There are actually two LEDs. Which by default, are set to a heartbeat pattern.
```
	blue_led@1 {
		gpios = <&gpio3 RK_PA6 GPIO_ACTIVE_HIGH>;
		label = "blue_led";
		linux,default-trigger = "heartbeat";
		linux,default-trigger-delay-ms = <0>;
	};
    
	green_led@2 {
		gpios = <&gpio3 RK_PB1 GPIO_ACTIVE_HIGH>;
		label = "green_led";
		linux,default-trigger = "heartbeat";
		linux,default-trigger-delay-ms = <0>;
	};
```
The LED trigger can be changed for both in the device tree or the exposed sysfs LED nodes.

**Reply by ethanc8** [reply]
@Joshua-Riek `/boot/firmware/ubuntuEnv.txt` no longer exists. Is there an alternate way to enable this device tree overlay?

**Reply by ethanc8** [reply]
Never mind, I found we can change `/etc/default/u-boot`:

```sh
## /etc/default/u-boot - configuration file for u-boot-update(8)

#U_BOOT_UPDATE="true"

#U_BOOT_ALTERNATIVES="default recovery"
#U_BOOT_DEFAULT="l0"
#U_BOOT_PROMPT="1"
#U_BOOT_ENTRIES="all"
#U_BOOT_MENU_LABEL="Debian GNU/Linux"
#U_BOOT_PARAMETERS="ro earlycon"
#U_BOOT_ROOT=""
#U_BOOT_TIMEOUT="50"
#U_BOOT_FDT=""
#U_BOOT_FDT_DIR="/lib/firmware/"
U_BOOT_FDT_OVERLAYS="device-tree/rockchip/overlay/orangepi-5-plus-disable-leds.dtbo"
#U_BOOT_FDT_OVERLAYS_DIR="/lib/firmware/"
#U_BOOT_SYNC_DTBS="false"
```

**Comment by SilverDesert** [comment]
Thanks but I have no idea how to make the leds blink meaningfully. I also have no idea why on earth they made the leds pulsate like a heartbeat.

**Reply by Joshua-Riek** [reply]
I believe the idea is, once the led starts to blink, that means the kernel has loaded and the OS is starting to boot.

**Reply by SilverDesert** [reply]
Thanks for your help.

**Comment by SilverDesert** [comment]
How can I make the led flash green instead of blue and vice versa?

**Reply by SilverDesert** [reply]
Is it possible?

**Reply by Joshua-Riek** [reply]
Yes, sorry I've had a busy week, I and give an example later today.

**Reply by SilverDesert** [reply]
How can I make the led flash green instead of blue and vice versa?
Thanks.

**Reply by Joshua-Riek** [reply]
Sorry about the delay. I forgot to get back to you. 

```bash
# Turn off the blue LED
echo none > /sys/class/leds/blue_led/trigger

# Turn off the green LED
echo none > /sys/class/leds/green_led/trigger

# Blink the blue LED in a heartbeat pattern
echo heartbeat > /sys/class/leds/blue_led/trigger

# Blink the green LED in a heartbeat pattern
echo heartbeat > /sys/class/leds/green_led/trigger
```

Look ledtrig in the kernel for more details.
https://www.kernel.org/doc/html/latest/leds/ledtrig-oneshot.html

**Reply by SilverDesert** [reply]
Thank you for your kind help.

**Comment by hvownij** [comment]
I have a Orange PI 5 Max.  The suggested elegant way does not work.  I have tried to swap the Plus out with Max and the command didnt work.

it seems they have changed to location of the LEDs or something over the years.


Only work around I found that actually works is to disable both the Blue and Green LED's with the following Commands in /etc/rc.local

You still get a few flashy heartbeats but as soon as its booted up and read the rc.local service it stops flashing and you get the steady RED led to stay on.

```
echo none > /sys/class/leds/green_led/trigger
echo none > /sys/class/leds/blue_led/trigger

**Reply by nobara256** [reply]
You need to change the device tree sources and remove the heartbeat trigger, recompile the device tree and load the generated ".dtb" or write a simple overlay, compile it with DTC and load it. It's a rabbit hole indeed, but I guess that's the only way of properly disabling it.

**Comment by tehmessiah75** [comment]
I can confirm that works for the Orange Pi 5 Pro too.
I had to change to super user though.
Running 
```
sudo echo none > /sys/class/leds/green_led/trigger
sudo echo none > /sys/class/leds/blue_led/trigger
```
Gave me this error.
```
sudo echo none > /sys/class/leds/blue_led/trigger 
bash: /sys/class/leds/blue_led/trigger: Permission denied
```
after I ran `sudo su` and then ran the commands it worked fine.
